### PR TITLE
PERF: Improved performance for StringMethods.encode/decode

### PR DESCRIPTION
**Pull request by Winand**
Reusing encoding/decoding function for a specified encoding doubles speed.

```
import pandas as pd
s1 = pd.Series(["Dzień добрый"]*100000)
s2 = s1.str.encode('cp1251', 'ignore')
%timeit s1.str.encode('cp1251', 'ignore')
%timeit s2.str.decode('cp1251')

```

10 loops, best of 3: 157 ms per loop --> 80.9 ms per loop
10 loops, best of 3: 179 ms per loop --> 84.1 ms per loop

### CODE REVIEW

**Comment by jreback** [comment]
pls add an asv benchmark for this

File: doc/source/whatsnew/v0.18.1.txt [hunk 1]
```diff
@@ -338,7 +338,7 @@ Performance Improvements
 
 
 - Improved performance of ``DataFrame.to_sql`` when checking case sensitivity for tables. Now only checks if table has been created correctly when table name is not lower case. (:issue:`12876`)
-
+- Improved performance of ``StringMethods.encode`` and ``decode`` methods
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: ``.str.encode()`` and ``.str.decode()``

**Comment by jreback** [comment]
so even better is to do this:

```
In [11]: %timeit s1.astype('category').str.decode('utf-8')
100 loops, best of 3: 4.39 ms per loop
```

This is of course is dependent on the number of uniques, so there is some overhead here if k (nuniques) is about n (len).

**Comment by Winand** [comment]
Tested 3 different encodings, this solution is better only in certain cases. Maybe this should be optimized in Python itself, not pandas.
```
import pandas as pd
s1 = pd.Series(["йцукенгшщз", "фывапролдж", "ячсмитьбю.", "ъхзщшгнеку", "эждлорпавы",
			  "окнсдветур", "ЙЦУКЕнгшщз", "ЭЖДЛОрпавы", "ЯЧСМИтьбю.", "ХЖЩЛГреауы"])
s1 = pd.concat([s1]*100000)
enc = 'utf-16'
s2 = s1.str.encode(enc)
%timeit s1.str.encode(enc)
%timeit s2.str.decode(enc)
enc = 'utf-8'
s2 = s1.str.encode(enc)
%timeit s1.str.encode(enc)
%timeit s2.str.decode(enc)
enc = 'cp1251'
s2 = s1.str.encode(enc)
%timeit s1.str.encode(enc)
%timeit s2.str.decode(enc)

1 loop, best of 3: 1.05 s per loop -> 437 ms (+) utf-16
1 loop, best of 3: 531 ms per loop -> 804 ms (-)
1 loop, best of 3: 454 ms per loop -> 483 ms (-) utf-8
1 loop, best of 3: 565 ms per loop -> 869 ms (-)
1 loop, best of 3: 1.51 s per loop -> 766 ms (+) cp1251
1 loop, best of 3: 1.76 s per loop -> 827 ms (+)
```

**Pull request closed without merging**